crates/interpreter/src/side_table.rs

```diff
@@ -12,14 +12,74 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use core::slice;
+
+use bytemuck::cast;
+
 use crate::bit_field::*;
 use crate::error::*;
+use crate::parser;
+use crate::toctou::Use;
+
+pub type Parser<'m> = parser::Parser<'m, Use>;
+
+#[allow(dead_code)]
+pub struct SideTable<'m> {
+    indices: &'m [u16], // including 0 and the length of metadata_array
+    metadata: &'m Metadata<'m>,
+}
+#[allow(dead_code)]
+impl<'m> SideTable<'m> {
+    fn metadata(&self, func_idx: usize) -> Metadata<'m> {
+        Metadata(
+            &self.metadata.0
+                [self.indices[func_idx] as usize .. self.indices[func_idx + 1] as usize],
+        )
+    }
+}
+
+#[allow(dead_code)]
+struct Metadata<'m>(&'m [u8]);
+
+#[allow(dead_code)]
+impl<'m> Metadata<'m> {
+    pub fn type_idx(&self) -> usize {
+        let bytes = &self.0[.. size_of::<usize>()];
```

Comment:
We shouldn't use 4 bytes for type idx, I expect 2 to be enough, so it's really just reading one element of the `[u16]`.